### PR TITLE
Implement recursive dataclass parsing

### LINKED ISSUE #15: Nested dataclasses
Support dataclasses as an argument.

For example:
```python
@dataclass
class Sub:
  a: int
  b: str

@dataclass
class Config:
  sub: Sub
```

It would be great if this could be supported with:

1. `--sub-a 42 --sub-b abc`, as well as
2. `--sub '{"a": 42, "b": "abc"}`

### PR DESCRIPTION
**Pull request by rogiervandergeer**
This allows fields of the dataclass to be other dataclasses.

This PR closes #15 and replaces #22

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -190,3 +191,47 @@ optional arguments:
...
+  --config-field-b CONFIG_FIELD_B
+                        (default: abc)
+
+```
+
+Please be aware of the following:
+- The default (factory) of fields with a dataclass type is ignored by pydargs, which may yield unexpected results.
+  If you must add a default, do `field_name: NestedClass = field(default_factory: NestedClass)`.
+- Nested dataclasses can not be positional (although _fields of_ the nested dataclass can be).
+- Argument names must not collide. In the example above, the `Base` class should not contain a field named `config_field_a`.
```
Comment (antonsteenvoorden) [inline README.md:239]: Maybe also write about the behaviour of what happens when this is done. 
Based on the tests it looks like this is not possible and will crash, but it's not immediately clear

File: tests/test_recursive.py [hunk 2]
```diff
@@ -0,0 +1,253 @@
...
+        "help_string",
+        [
+            "usage: prog [-h] [--sub-a SUB_A] [--sub-b-a SUB_B_A]",
+            "sub_b:",
+            "--sub-b-b SUB_B_B  a string (default: abc)\n",
+        ],
+    )
+    def test_help(self, capsys, help_string: str):
+        with raises(SystemExit):
+            parse(self.Config, ["--help"], prog="prog")  # type: ignore
```
Comment (antonsteenvoorden) [inline tests/test_recursive.py:222]: why does this raise a system exit?

File: src/pydargs/__init__.py [hunk 3]
```diff
@@ -45,11 +45,31 @@ def parse(tp: Type[Dataclass], args: Optional[list[str]] = None, **kwargs: Any)
         An instance of tp.
     """
     namespace = _create_parser(tp, **kwargs).parse_args(args)
-    return tp(**namespace.__dict__)
+    return _create_object(tp, namespace)
+
+
+def _create_object(tp: Type[Dataclass], namespace: Namespace, prefix: str = "") -> Dataclass:
+    for field in fields(tp):
+        if hasattr(field.type, "__dataclass_fields__"):
```
Comment (antonsteenvoorden) [inline src/pydargs/__init__.py:55]:
```suggestion
        if is_dataclass := hasattr(field.type, "__dataclass_fields__"):
```

An idea to indicate what we are checking here

File: src/pydargs/__init__.py [hunk 4]
```diff
@@ -45,11 +45,31 @@ def parse(tp: Type[Dataclass], args: Optional[list[str]] = None, **kwargs: Any)
...
 
 def _create_parser(tp: Type[Dataclass], **kwargs: Any) -> ArgumentParser:
-    parser = ArgumentParser(**kwargs, argument_default=argparse.SUPPRESS)
+    parser = ArgumentParser(**kwargs, argument_default=SUPPRESS)
+    _add_arguments(parser, tp)
+    return parser
+
+
+def _add_arguments(parser: ArgumentParser, tp: Type[Dataclass], prefix: str = "") -> ArgumentParser:
+    parser_or_group = parser.add_argument_group(prefix.strip("_")) if len(prefix) else parser
```
Comment (antonsteenvoorden) [inline src/pydargs/__init__.py]:
```suggestion
    parser_or_group = parser.add_argument_group(prefix.strip("_")) if prefix else parser
```

File: src/pydargs/__init__.py [hunk 5]
```diff
@@ -111,15 +131,22 @@ def _create_parser(tp: Type[Dataclass], **kwargs: Any) -> ArgumentParser:
...
                 setattr(union_parser, "__name__", repr(field.type))
-                parser.add_argument(
+                parser_or_group.add_argument(
                     *arguments,
                     type=union_parser,
                     **argument_kwargs,
                 )
             else:
                 raise NotImplementedError(f"Parsing into type {origin} is not implemented.")
+        elif hasattr(field.type, "__dataclass_fields__"):
```
Comment (antonsteenvoorden) [inline src/pydargs/__init__.py:143]:
```suggestion
        elif is_dataclass := hasattr(field.type, "__dataclass_fields__"):
```

File: src/pydargs/__init__.py [hunk 6]
```diff
@@ -111,15 +131,22 @@ def _create_parser(tp: Type[Dataclass], **kwargs: Any) -> ArgumentParser:
...
+            if positional:
+                raise ValueError("Dataclasses may not be positional arguments.")
+            if field.default_factory is not None and field.default_factory != field.type:
+                warn(f"Non-standard default of field {field.name} is ignored by pydargs.", UserWarning)
```
Comment (antonsteenvoorden) [inline src/pydargs/__init__.py:147]: This refers to the default provided in e.g. SubConfig(a=1) right? 
```suggestion
                warn(f"Non-standard default for nested dataclass {field.name} is ignored by pydargs.", UserWarning)
```

File: src/pydargs/__init__.py [hunk 7]
```diff
@@ -111,15 +131,22 @@ def _create_parser(tp: Type[Dataclass], **kwargs: Any) -> ArgumentParser:
...
+            _add_arguments(parser, field.type, prefix=f"{prefix}{field.name}_")
```
Comment (antonsteenvoorden) [inline src/pydargs/__init__.py:149]:
```suggestion
            # Recursively add arguments for the nested dataclasses
            _add_arguments(parser, field.type, prefix=f"{prefix}{field.name}_")
```
I think this makes it easier to spot the recursion in the code?

File: src/pydargs/__init__.py [hunk 8]
```diff
@@ -111,15 +131,22 @@ def _create_parser(tp: Type[Dataclass], **kwargs: Any) -> ArgumentParser:
...
+            argument_kwargs["required"] = False
```
Comment (antonsteenvoorden) [inline src/pydargs/__init__.py]: why is this necessary? And shouldn't we completely remove this argument from the parser? We don't want the "parent" argument of the nested subclass as part of the cli right? because we can't provide a full dataclass via the CLI, but only build it up from the nested arguments?

File: tests/test_recursive.py [hunk 9]
```diff
@@ -0,0 +1,253 @@
...
+class TestSubConfigCollision:
+    @dataclass
+    class Config:
+        mode: str = field(metadata=dict(positional=True))
+        sub: SubConfig = field(default_factory=SubConfig)
+        sub_a: float = field(default=1.0)
+
+    def test_parse(self):
+        with raises(ArgumentError):
+            parse(self.Config, ["mode"])
```
Comment (antonsteenvoorden) [inline tests/test_recursive.py:83]: is this argparse raising an error when duplicate keys are added?

File: README.md [hunk 10]
```diff
@@ -190,3 +191,47 @@ optional arguments:
...
+config:
+  --config-field-a CONFIG_FIELD_A
+  --config-field-b CONFIG_FIELD_B
+                        (default: abc)
+
+```
+
+Please be aware of the following:
+- The default (factory) of fields with a dataclass type is ignored by pydargs, which may yield unexpected results.
+  If you must add a default, do `field_name: NestedClass = field(default_factory: NestedClass)`.
```
Comment (antonsteenvoorden) [inline README.md]:
```suggestion
- The default (factory) of fields with a dataclass type is ignored by pydargs, which may yield unexpected results. 
  I.e. in the provided default for `a` in `field_name: NestedClass = field(default_factory: NestedClass(a=1))` is ignored. 
  If you want to set a defaults for the NestedClass, you must do this in the class definition of `NestedClass`
  and set the default for the nested class as `field_name: NestedClass = field(default_factory: NestedClass)`.
```

Comment (antonsteenvoorden) [reply README.md]: This is a little more explicit, it took me some time to figure out what you meant

File: src/pydargs/__init__.py [hunk 11]
```diff
@@ -45,11 +45,31 @@ def parse(tp: Type[Dataclass], args: Optional[list[str]] = None, **kwargs: Any)
...
+    for field in fields(tp):
+        if hasattr(field.type, "__dataclass_fields__"):
+            setattr(
+                namespace,
+                prefix + field.name,
+                _create_object(field.type, namespace, prefix=f"{prefix}{field.name}_"),
+            )
+    args = {key[len(prefix) :]: value for key, value in namespace.__dict__.items() if key.startswith(prefix)}
+    for key in args.keys():
+        delattr(namespace, prefix + key)
```
Comment (antonsteenvoorden) [inline src/pydargs/__init__.py]: Can you add a little docstring here that explains what's happening?

File: tests/test_recursive.py [hunk 12]
```diff
@@ -0,0 +1,253 @@
...
+        assert config.sub.a == 1
+        assert config.sub.b == "abc"
+        assert config.flag is False
+
+
+class TestParseSubConfigWithRequiredField:
+    @dataclass
+    class Config:
+        mode: str = field(metadata=dict(positional=True))
+        sub: SubConfigWithRequiredField
```
Comment (antonsteenvoorden) [inline tests/test_recursive.py:131]: I'm surprised a default factory is not needed here 🤔

File: README.md (hunk 10 above)
Comment (antonsteenvoorden) [reply README.md]: Is this last line also correct? or is the default factory not required? if not, why would a default_factory be needed? For the ordering of the properties on the class? i.e. properties without defaults after properties with defaults?

File: tests/test_recursive.py [hunk 13]
```diff
@@ -0,0 +1,253 @@
...
+
+
+class TestParsePositionalSubConfig:
+    @dataclass
+    class Config:
+        mode: str = field(metadata=dict(positional=True))
+        sub: SubConfig = field(default_factory=SubConfig, metadata=dict(positional=True))
+        flag: bool = field(default=False, metadata=dict(as_flags=True))
+
+    def test_parse(self):
```
Comment (antonsteenvoorden) [inline tests/test_recursive.py:200]: is it possible to add a capsys check here?

**Review by antonsteenvoorden — CHANGES_REQUESTED** [review]
Nicely done, lefts some minor comments and questions

File: tests/test_recursive.py (hunk 2 above)
Comment (rogiervandergeer) [reply tests/test_recursive.py:222]: When calling `--help`, argparse always prints a help message and then exits.

File: tests/test_recursive.py (hunk 9 above)
Comment (rogiervandergeer) [reply tests/test_recursive.py:83]: Exactly! We need to verify that this already fails when creating the parser, and not (only) when we actually use the arguments.

File: src/pydargs/__init__.py (hunk 3 above)
Comment (rogiervandergeer) [reply src/pydargs/__init__.py:55]: That isn't allowed by ruff;  ```F841 Local variable `is_dataclass` is assigned to but never used```

File: src/pydargs/__init__.py (hunk 8 above)
Comment (rogiervandergeer) [reply src/pydargs/__init__.py]: It isn't. I removed it. It was a remnant of an attempt to allow json to configure a nested dataclass `--sub {"a": 1}`. I've abandoned that attempt for now.

File: tests/test_recursive.py (hunk 12 above)
Comment (rogiervandergeer) [reply tests/test_recursive.py:131]: Defaults are only necessary when field defined prior have a default. The `mode` field has no default

File: tests/test_recursive.py (hunk 13 above)
Comment (rogiervandergeer) [reply tests/test_recursive.py:200]: Sure - but there will be no output. This raises an exception, which differs from the `SystemExit` raised by argparse after it prints a helpful message.

File: src/pydargs/__init__.py (hunk 6 above)
Comment (rogiervandergeer) [reply src/pydargs/__init__.py:147]: It does, indeed. I don't agree that this suggestion improves the warning; for the user the "nested" may not be clear. In any case the dataclass in question is just a field of the dataclass the user is parsing, so that should be clear.

File: README.md (hunk 1 above)
Comment (rogiervandergeer) [reply README.md:239]: In that case you'll get an `ArgumentError` indeed. Does the user need to know that? They shouldn't get in that situation anyway.

Comment (antonsteenvoorden) [reply README.md:239]:
```suggestion
- Argument names must not collide. In the example above, the `Base` class cannot contain a field named `config_field_a`.
```

File: src/pydargs/__init__.py (hunk 3 above)
Comment (antonsteenvoorden) [reply src/pydargs/__init__.py:55]: Ah crap..

File: src/pydargs/__init__.py [hunk 14]
```diff
@@ -45,11 +45,32 @@ def parse(tp: Type[Dataclass], args: Optional[list[str]] = None, **kwargs: Any)
...
+    for field in fields(tp):
+        if hasattr(field.type, "__dataclass_fields__"):
+            # Create nested dataclass object
+            setattr(
+                namespace,
+                prefix + field.name,
+                _create_object(field.type, namespace, prefix=f"{prefix}{field.name}_"),
+            )
+    args = {key[len(prefix) :]: value for key, value in namespace.__dict__.items() if key.startswith(prefix)}
+    for key in args.keys():  # Remove used keys from the namespace
```
Comment (antonsteenvoorden) [inline src/pydargs/__init__.py]: What do you mean by "used keys"?

File: tests/test_recursive.py (hunk 12 above)
Comment (antonsteenvoorden) [reply tests/test_recursive.py:131]: Yeah but also for lists etc. I expected this to fall under the same category

File: README.md (hunk 10 above)
Comment (rogiervandergeer) [reply README.md]: Your suggestion isn't entirely correct. The note isn't about arguments to a default constructor, but about the entire default and/or default_factory. Whatever is provided as `default` or `default_factory`, is completely ignored.

Would this be better?

```suggestion
- The default (factory) of fields with a dataclass type is ignored by pydargs, which may yield unexpected results.
  E.g., in the example above, `config: Config = field(default_factory=lambda: Config(field_b="def"))` will not result in a default of "def" for field_b when parsed by pydargs.
  If you must add a default, for example for instantiating your dataclass elsewhere, do `config: Config = field(default_factory=Config)`.
```

File: tests/test_recursive.py (hunk 13 above)
Comment (antonsteenvoorden) [reply tests/test_recursive.py:200]: Ah okay that's less useful indeed

File: src/pydargs/__init__.py (hunk 14 above)
Comment (rogiervandergeer) [reply src/pydargs/__init__.py]: The keys that have been used to instantiate an object.

For example, in the example at the end of the readme, the (dict repr of the) namespace could look like:
`{"verbose": "yes", "config_field_a": 12, "config_field_b": "def"}`.

 First `Config` is instantiated using all keys that start with `"config_"` (and with the prefix removed), and added to the namespace:
 `config=Config(field_a=12, field_b="def")`.

Then we instantiate `Base`, with all keys that start with `""`, i.e. all keys. This requires that `config_field_a` and `config_field_b` have been removed.

File: README.md (hunk 1 above)
Comment (rogiervandergeer) [reply README.md:239]: I deliberately chose the wording "should not" as it is not recommended, but one could in principle have a field named `config_field_a` with `ignore_arg=True` to prevent collisions.

See [RFC2119](https://www.rfc-editor.org/rfc/pdfrfc/rfc2119.txt.pdf).

File: README.md (hunk 10 above)
Comment (antonsteenvoorden) [reply README.md]: > If you want to set a defaults for the NestedClass, you must do this in the class definition of `NestedClass`

This part of my suggestion is still relevant though, right? 

> If you must add a default, for example for instantiating your dataclass elsewhere, do `config: Config = field(default_factory=Config)`.

Is this strictly required when instantiation the object manually? 

If so, I think we can rephrase it as:
> To support instantiating the dataclass directly, you should do `config: Config = field(default_factory=Config)`.

File: README.md (hunk 1 above)
Comment (antonsteenvoorden) [reply README.md:239]: Ahhhhh

Comment (antonsteenvoorden) [reply README.md:239]: I think this is a good example to add to the docstring though, I didn't think of that

Comment (antonsteenvoorden) [reply README.md:239]: Something like: "It is possible, when setting `ignore_arg` to True, but not recommended."

File: src/pydargs/__init__.py (hunk 14 above)
Comment (antonsteenvoorden) [inline src/pydargs/__init__.py]:
```suggestion
    # Remove prefixed keys from the namespace after instantiating nested object.
    for key in args.keys():
```

Comment (antonsteenvoorden) [reply src/pydargs/__init__.py]: 👍

File: README.md (hunk 10 above)
Comment (rogiervandergeer) [reply README.md]: > This part of my suggestion is still relevant though, right?

True; how about this?

```suggestion
- The default (factory) of fields with a dataclass type is ignored by pydargs, which may yield unexpected results.
  E.g., in the example above, `config: Config = field(default_factory=lambda: Config(field_b="def"))` will not result in a default of "def" for field_b when parsed by pydargs. 
  Instead, set `field_b: str = "def"` in the definition of `Config`.
  If you must add a default, for example for instantiating your dataclass elsewhere, do `config: Config = field(default_factory=Config)`, assuming that all fields in `Config` have a default.
```

Having a default for a dataclass-type field is not necessary for instantiating it elsewhere, but it is required if you want to be able to instantiate it without arguments.

File: src/pydargs/__init__.py (hunk 14 above)
Comment (rogiervandergeer) [reply src/pydargs/__init__.py]: Note that the keys that are removed are not necessarily prefixed, nor for nested objects.

File: README.md (hunk 1 above)
Comment (rogiervandergeer) [reply README.md:239]: Do we really want to go into such detail? The wording "should not" should tell the user that in principle it is a bad idea, unless they understand the full implications of doing so. For the purpose of a README that should be enough, right? This is something that no-one will likely encounter anyway.

Comment (antonsteenvoorden) [reply README.md:239]: 👍

File: README.md (hunk 10 above)
Comment (antonsteenvoorden) [reply README.md]: Yes that looks good! Note that on line `#236` you have two dots. Not sure if intentional

File: src/pydargs/__init__.py (hunk 14 above)
Comment (antonsteenvoorden) [reply src/pydargs/__init__.py]: Oh dip, it's _that_ generic! Nice. 
However, if possible, I would like to make explicit that this is mostly required for those cases (it is, right? 🤔).
If I understand this correctly: If there are no nested dataclasses, there not going to be a call to this method with a prefix. Meaning this won't recurse and the namespace is cleared but for no reason. So this functionality does purely exist to support that use case...

... and it's not super apparent (to me)

Comment (rogiervandergeer) [reply src/pydargs/__init__.py]: How about:

```suggestion
    # Remove the keys used so far from the namespace, to prevent clutter when creating a parent object.
    for key in args.keys():
```

Comment (antonsteenvoorden) [reply src/pydargs/__init__.py]: :+1:

**Review by antonsteenvoorden — APPROVED** [review]
👍  Nice addition

**Comment by rogiervandergeer** [comment]
Thanks for the review!
